analyzers/src/SonarAnalyzer.CSharp/Rules/SillyMathematicalComparison.cs

```diff
@@ -0,0 +1,116 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class SillyMathematicalComparison : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S2198";
+        private const string MathComparisonMessage = "Comparison to this constant is useless; the constant is outside the range of type '{0}'";
+
+        private static readonly DiagnosticDescriptor MathComparisonRule = DescriptorFactory.Create(DiagnosticId, MathComparisonMessage);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(MathComparisonRule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterNodeAction(
+                CheckBinary,
+                SyntaxKind.GreaterThanExpression,
+                SyntaxKind.GreaterThanOrEqualExpression,
+                SyntaxKind.LessThanExpression,
+                SyntaxKind.LessThanOrEqualExpression,
+                SyntaxKind.EqualsExpression,
+                SyntaxKind.NotEqualsExpression);
+
+        private static void CheckBinary(SonarSyntaxNodeReportingContext context)
+        {
+            if (TryGetConstantValue(context, (BinaryExpressionSyntax)context.Node, out var constant, out var other)
+               && context.SemanticModel.GetTypeInfo(other).Type is { } typeSymbolOfOther
+               && TryGetRange(typeSymbolOfOther, out var min, out var max)
+               && (constant < min || constant > max))
+            {
+                var typeName = context.SemanticModel.GetTypeInfo(other).Type.ToMinimalDisplayString(context.SemanticModel, other.GetLocation().SourceSpan.Start);
+                context.ReportIssue(Diagnostic.Create(MathComparisonRule, other.Parent.GetLocation(), typeName));
+            }
+        }
+
+        private static bool TryGetConstantValue(SonarSyntaxNodeReportingContext context, BinaryExpressionSyntax binary, out double constant, out SyntaxNode other)
+        {
+            constant = default;
+            other = default;
+            var maybeLeft = context.SemanticModel.GetConstantValue(binary.Left);
+            var maybeRight = context.SemanticModel.GetConstantValue(binary.Right);
+
+            if (maybeLeft.HasValue ^ maybeRight.HasValue)
+            {
+                if (maybeLeft.HasValue)
+                {
+                    other = binary.Right;
+                    return TryConvertToDouble(maybeLeft.Value, out constant);
+                }
+                else
+                {
+                    other = binary.Left;
+                    return TryConvertToDouble(maybeRight.Value, out constant);
+                }
```

Comment:
It's more sound to set all out parameters only if successful.
```suggestion
                if (maybeLeft.HasValue && TryConvertToDouble(maybeLeft.Value, out constant))
                {
                    other = binary.Right;
                    return true;
                }
                else if (TryConvertToDouble(maybeRight.Value, out constant);
                {
                    other = binary.Left;
                    return true;
                }
```